File: clients/reactive-dart/lib/state_store.dart
/// Synchronous state management without Flutter, I/O, or a second sync queue.
library;

/// Reducers, states and selected values must be immutable. Effects live outside
/// the reducer and enter the store through a [StateEffect].
final class StateStore<S, A> {
  StateStore(
    S initial,
    S Function(S state, A action) reduce, {
    void Function(Object error, StackTrace stackTrace)? onObserverError,
  }) : _state = initial,
       _reduce = reduce,
       _onObserverError = onObserverError;

  S _state;
  final S Function(S, A) _reduce;
  final void Function(Object, StackTrace)? _onObserverError;
  var _revision = 0;
  var _busy = false;
  var _disposed = false;
  final _listeners = <void Function()>{};
  final _effects = <String, Object>{};

  S get state => _state;
  int get revision => _revision;
  bool get disposed => _disposed;

  void _check() {
    if (_disposed) throw StateError('StateStore is disposed');
    if (_busy) {
      throw StateError('StateStore does not allow reentrant transitions');
    }
  }

  void _notify() {
    for (final listener in List<void Function()>.of(_listeners)) {
      if (!_listeners.contains(listener)) continue;
      try {
        listener();
      } catch (error, stackTrace) {
        try {
          _onObserverError?.call(error, stackTrace);
        } catch (_) {
          // A rendering/diagnostic failure cannot fail a committed dispatch.
        }
      }
    }
  }

  void dispatch(A action) {
    _check();
    _busy = true;
    try {
      final next = _reduce(_state, action);
      if (next is Future) {
        throw StateError('StateStore reducers must be synchronous');
      }
      _state = next;
      _revision += 1;
      _notify();
    } finally {
      _busy = false;
    }
  }

  /// Clear user state and fence outstanding results on session change/logout.
  void reset(S initial) {
    _check();
    _busy = true;
    try {
      _effects.clear();
      _state = initial;
      _revision += 1;
      _notify();
    } finally {
      _busy = false;
    }
  }

  /// Replay synchronously, then deliver only distinct selected values.
  /// The returned cancellation function is idempotent.
  void Function() select<T>(
    T Function(S state) selector,
    void Function(T value) listener, {
    bool Function(T left, T right)? same,
  }) {
    _check();
    _busy = true;
    try {
      final equals = same ?? (T left, T right) => left == right;
      var previous = selector(_state);
      void notify() {
        final next = selector(_state);
        if (equals(previous, next)) return;
        previous = next;
        listener(next);
      }

      listener(previous);
      _listeners.add(notify);
      return () => _listeners.remove(notify);
    } finally {
      _busy = false;
    }
  }

  /// Latest result wins per key. Does not cancel already-started durable writes.
  StateEffect<A> beginEffect(String key) {
    _check();
    final token = Object();
    _effects[key] = token;
    bool isCurrent() => !_disposed && identical(_effects[key], token);
    return StateEffect<A>._(
      isCurrent,
      (action) {
        if (!isCurrent()) return false;
        dispatch(action);
        return true;
      },
      () {
        if (isCurrent()) _effects.remove(key);
      },
    );
  }

  /// Hydrate a complete localView after pending mutations have been replayed.
  Future<bool> projectLocalView<T>(
    String key,
    Future<T> Function() read,
    A Function(T value) action,
  ) async {
    final effect = beginEffect(key);
    try {
      final value = await read();
      return effect.isCurrent && effect.dispatch(action(value));
    } finally {
      effect.close();
    }
  }

  void dispose() {
    if (_disposed) return;
    _check();
    _disposed = true;
    _effects.clear();
    _listeners.clear();
  }
}

final class StateEffect<A> {
  StateEffect._(this._isCurrent, this.dispatch, this.close);

  final bool Function() _isCurrent;
  final bool Function(A action) dispatch;
  final void Function() close;
  bool get isCurrent => _isCurrent();
}

File: clients/reactive-ts/src/state-store.ts
/** A synchronous, UI-independent store. State and selected values must be immutable. */
export class StateStore<S, A> {
  #state: S;
  #revision = 0;
  #busy = false;
  #disposed = false;
  #listeners = new Set<() => void>();
  #effects = new Map<string, object>();
  readonly #reduce: (state: Readonly<S>, action: A) => S;
  readonly #onObserverError: (error: unknown) => void;

  constructor(initial: S, reduce: (state: Readonly<S>, action: A) => S,
    onObserverError: (error: unknown) => void = () => {}) {
    this.#state = initial;
    this.#reduce = reduce;
    this.#onObserverError = onObserverError;
  }

  get state(): Readonly<S> { return this.#state; }
  get revision(): number { return this.#revision; }
  get disposed(): boolean { return this.#disposed; }

  #check(): void {
    if (this.#disposed) throw new Error('StateStore is disposed');
    if (this.#busy) throw new Error('StateStore does not allow reentrant transitions');
  }

  #notify(): void {
    for (const listener of [...this.#listeners]) {
      if (!this.#listeners.has(listener)) continue;
      try { listener(); } catch (error) {
        // Rendering/diagnostics must not turn a committed dispatch into a failure.
        try { this.#onObserverError(error); } catch { /* observer isolation */ }
      }
    }
  }

  /** Reducer failures leave the previous state and revision intact. */
  dispatch(action: A): void {
    this.#check();
    this.#busy = true;
    try {
      const next = this.#reduce(this.#state, action);
      if (next !== null && typeof next === 'object' &&
          'then' in next && typeof next.then === 'function') {
        throw new TypeError('StateStore reducers must be synchronous');
      }
      this.#state = next;
      this.#revision += 1;
      this.#notify();
    } finally { this.#busy = false; }
  }

  /** Clear user state and invalidate every outstanding effect on session change. */
  reset(initial: S): void {
    this.#check();
    this.#busy = true;
    try {
      this.#effects.clear();
      this.#state = initial;
      this.#revision += 1;
      this.#notify();
    } finally { this.#busy = false; }
  }

  /** Replays once synchronously, then emits only when the selected value changes. */
  select<T>(selector: (state: Readonly<S>) => T, listener: (value: T) => void,
    same: (left: T, right: T) => boolean = Object.is): () => void {
    this.#check();
    this.#busy = true;
    try {
      let previous = selector(this.#state);
      const notify = () => {
        const next = selector(this.#state);
        if (same(previous, next)) return;
        previous = next;
        listener(next);
      };
      // Do not retain the listener if initial selection/delivery fails.
      listener(previous);
      this.#listeners.add(notify);
      return () => { this.#listeners.delete(notify); };
    } finally { this.#busy = false; }
  }

  /** Latest result wins per key; unrelated keys can run concurrently. */
  beginEffect(key: string): StateEffect<A> {
    this.#check();
    const token = {};
    this.#effects.set(key, token);
    const isCurrent = () => !this.#disposed && this.#effects.get(key) === token;
    return {
      isCurrent,
      dispatch: (action) => {
        if (!isCurrent()) return false;
        this.dispatch(action);
        return true;
      },
      close: () => { if (isCurrent()) this.#effects.delete(key); },
    };
  }

  /** Read a complete, already-rebased localView; never hydrate a raw remote echo. */
  async projectLocalView<T>(key: string, read: () => Promise<T>, action: (value: T) => A): Promise<boolean> {
    const effect = this.beginEffect(key);
    try {
      const value = await read();
      return effect.isCurrent() && effect.dispatch(action(value));
    } finally { effect.close(); }
  }

  /** Releases subscriptions and invalidates effects. Does not touch durable queues. */
  dispose(): void {
    if (this.#disposed) return;
    this.#check();
    this.#disposed = true;
    this.#effects.clear();
    this.#listeners.clear();
  }
}

/** Logical cancellation only: an already-started durable write must still settle. */
export interface StateEffect<A> {
  isCurrent(): boolean;
  dispatch(action: A): boolean;
  close(): void;
}
